text/2017-11-09-channel.md

```diff
@@ -0,0 +1,944 @@
+
+## Design goals
+
+All shortcomings of `std::sync::mpsc` enumerated in the *Motivation* section must be fixed.
+`crossbeam-channel` strives to look simple and stupid on the outside,
+but be ambitiously powerful inside. The channel will:
+
+1. Be an MPMC channel (allow multiple `Sender`s and multiple `Receiver`s).
+2. Avoid locks and be much faster than `std::sync::mpsc` in the bounded case (the `sync_channel` case).
+3. Not be significantly slower than `std::sync::mpsc` in the unbounded case.
+4. Have the full spectrum of `try_send`/`send`/`send_timeout`/`try_recv`/`recv`/`recv_timeout` methods.
+5. Have convenience methods like `capacity`, `len`, and `is_empty`.
+6. Have `Senders` and `Receivers` that implement `Sync`.
+7. Have a simple interface without special types like `SyncSender`.
+8. Support selection over both send and receive operations.
+9. Allow selection with a timeout.
+10. Have easy and safe dynamic selection.
+11. Have selection without the warts of `select!`.
+12. Have fair(er) selection.
+
+## The interface
+
+Constructors:
+
+```rust
+pub fn unbounded<T>() -> (Sender<T>, Receiver<T>);
+pub fn bounded<T>(cap: usize) -> (Sender<T>, Receiver<T>);
+```
+
+Sender:
+
+```rust
+pub struct Sender<T> { ... }
+
+unsafe impl<T: Send> Send for Sender<T> {}
+unsafe impl<T: Send> Sync for Sender<T> {}
+
+impl<T> Sender<T> {
+    pub fn try_send(&self, value: T) -> Result<(), TrySendError<T>>;
+    pub fn send(&self, value: T) -> Result<(), SendError<T>>;
+    pub fn send_timeout(&self, value: T, dur: Duration) -> Result<(), SendTimeoutError<T>>;
+
+    pub fn is_empty(&self) -> bool;
+    pub fn len(&self) -> usize;
+    pub fn capacity(&self) -> Option<usize>;
+    pub fn is_disconnected(&self) -> bool;
+}
+
+impl<T> Clone for Sender<T> { ... }
+```
+
+Receiver:
+
+```rust
+pub struct Receiver<T> { ... }
+
+unsafe impl<T: Send> Send for Receiver<T> {}
+unsafe impl<T: Send> Sync for Receiver<T> {}
+
+impl<T> Receiver<T> {
+    pub fn try_recv(&self) -> Result<T, TryRecvError>;
+    pub fn recv(&self) -> Result<T, RecvError>;
+    pub fn recv_timeout(&self, dur: Duration) -> Result<T, RecvTimeoutError>;
+
+    pub fn is_empty(&self) -> bool;
+    pub fn len(&self) -> usize;
+    pub fn capacity(&self) -> Option<usize>;
+    pub fn is_disconnected(&self) -> bool;
+
+    pub fn iter(&self) -> Iter<T>;
+    pub fn try_iter(&self) -> TryIter<T>;
+}
+
+impl<T> Clone for Receiver<T> { ... }
+```
+
+Iterators:
+
+```rust
+impl<'a, T> IntoIterator for &'a Receiver<T> { ... }
+impl<T> IntoIterator for Receiver<T> { ... }
+
+pub struct Iter<'a, T: 'a> { ... }
+impl<'a, T> Iterator for Iter<'a, T> { ... }
+
+pub struct TryIter<'a, T: 'a> { ... }
+impl<'a, T> Iterator for TryIter<'a, T> { ... }
+
+pub struct IntoIter<T> { ... }
+impl<T> Iterator for IntoIter<T> { ... }
+```
+
+Selection:
+
+```rust
+pub mod select {
+    pub fn send<T>(tx: &Sender<T>, value: T) -> Result<(), T>;
+    pub fn recv<T>(rx: &Receiver<T>) -> Result<T, ()>;
```

Comment:
Super minor nit, but I'd use a ZST error type like `struct NotReady;` or something here, rather than `()`.

Reply:
I like the symmetry between `Result<(), T>` and `Result<T, ()>`.

If `recv` returns a `Result<T, SelectRecvError>`, then `send` should probably similarly return a `Result<(), SelectSendError<T>>`. That means we'd have to regain ownership from a failed `send` operation like this:

```rust
if let Err(err) = sel.send(&tx, name) {
    // Sending failed. Regain ownership of the message.
    name = err.into_inner();
} else {
    // Success! The message was sent.
    break;
}
```

Calling `.into_inner()` is slightly more unergonomic, but it's not too bad. Possible alternatives could be `.into()` and `.take()`. Do you agree with this reasoning? :)